### ISSUE TITLE
[ServiceWorker] localhost took too long to respond. HTTP ERROR 504

### ISSUE THREAD

**Issue by MrCroft** [angular/angular#21636]
## I'm submitting a...
```
[x] Bug report
```

## Minimal reproduction of the problem with instructions
1. `ng new sw --service-worker --routing`
2. `cd sw`
3. `ng g c home`
4. `ng g c about`
5. Edit `app-routing.module.ts` to add the routes for the 2 components we've just created:
```
// ...
import { HomeComponent } from './home/home.component';
import { AboutComponent } from './about/about.component';

const routes: Routes = [
  { path: '', component: HomeComponent },
  { path: 'about', component: AboutComponent }
];

// ...
```
6. `ng build --prod`
7. `npm install -g http-server` (you can use any server to test this, if you already prefer another one)
8. `http-server dist`
9. Access `localhost:8080` (8080 is the default port that http-server uses at the time of reporting this issue)
10. Access `localhost:8080/about`
11. Stop the server
12. Now try reloading the app while on the `/about` route (if it works the first time, try refreshing a few more times or wait a little before trying again). The browser will display the message:
```
This page isn’t working
localhost took too long to respond.
HTTP ERROR 504
```
(With Universal/SSR, it stops working right away, most of the time with the first refresh after stopping the server, no matter the route - and the issue is not the hash in the generated ngsw.json, because I regenerate that at the end of the build process).

Now, if you start the server again, loading the app on the `/about` route works (even though http-server is not SPA friendly, it doesn't redirect all requests to index.html), but the service worker kicks in.
Initially I noticed this behavior on a Universal app, doing server side rendering (following the [universal-starter repo](https://github.com/angular/universal-starter)). So this is not a http-server "not being SPA friendly" issue.
I only used http-server to reproduce it because I wanted to keep the reproduction process as fast as possible, without introducing server side rendering.
The CLI version used (1.6.4) initializes the app with Angular 5.1.0, but I've also tested on 5.2.1. I've skipped mentioning this step above because it happens regardless of the Angular version (at least on `>= 5.1.0`, haven't tested with older versions).

## Environment

```
Angular CLI: 1.6.4
Node: 8.9.4
OS: win32 x64
Angular: 5.2.1

Browser:
I've only tested on:
- [x] Chrome (desktop) version 63.0.3239.132
- [x] Chrome (desktop) version 65.0.3323.2 (canary)
```

**Comment by MrCroft** [comment]
I've just realised that the same things happen with API requests (urls which are set up in "dataGroups"). If I stop the API server, I get `504 Gateway Timeout (from ServiceWorker)` if I open to check the network tab.

**Comment by odahcam** [comment]
I'm with the same error here, when I'm online I can see a lot of things being loaded by the service worker, but when I turn into online, nothing is loaded.

**Comment by odahcam** [comment]
Seems like when I disable the SW registration already having it registered, the offline functionality works fine. 🤔

**Comment by KGALLET** [comment]
Seems like I've got the same issue

**Comment by jackkoppa** [comment]
I believe the root cause would be the same as the issue for this SO question I posted: [How to handle routing in Angular 5+ Service Workers?](https://stackoverflow.com/questions/48565629/how-to-handle-routing-in-angular-5-service-workers) It doesn't seem that Angular 5+ gives us a way to route service worker handling for non-index routes, at least that I can find.

**Comment by DominicBoettger** [comment]
same error here? So is the service worker in ng5 not usable?

**Comment by odahcam** [comment]
In CLI projects it works fine, I really don't know what's the problem. 😕

**Comment by jackkoppa** [comment]
@odahcam - does that mean, in a CLI project on Angular 5, you're able to handle routing with the generated service worker? Sorry, not sure if you mean "you don't know where the problem comes from" or "you're not sure what problem other people are having, because it works for you"

**Comment by odahcam** [comment]
I don't know where the problem came from, I only saw it after `ng eject` and working without the CLI. With the CLI everything always worked fine for me. 😕 ²

**Comment by ajithere** [comment]
Getting exact same 504 error with API request setup in datagroups. Because of this while my audit mentions everything to be perfect with respect to PWA service worker, i am unable to run my application in offline mode. Waiting

**Comment by jackkoppa** [comment]
During testing of the issue, I've recreated the exact problem explained by @MrCroft in a repro [here](https://github.com/jackkoppa/service-worker-test/tree/routing-failure-minimal-repro), to ensure that it exists even with the simplest app from `ng new app-name --service-worker`. Still trying to dig into where the fetching fails in `ngsw-worker.js`.

**Comment by DominicBoettger** [comment]
My app loads a json file from assets on startup (lang pack). Can maybe xmlhttprequests on init cause this?

**Comment by jackkoppa** [comment]
I've been able to figure out two issue spots, that when worked around in my app, allow it to work offline. I won't go into full detail here; I've given a long explanation on [this SO answer](https://stackoverflow.com/a/48732922/4167438).

1. Hosted versions of an Angular app, where the --base-href flag is set from the CLI, list absolute URLs for their service worker resources. When comparing requests to these resources, ngsw-worker.js is expecting relative URLs
2. Of the 3 available "states" that the ngsw-worker.js can be in, in my experience, EXISTING_CLIENTS_ONLY seems to bet set incorrectly.

(source code links are in the SO answer; I can duplicate them here as well)

**My current workaround is running [this build script](https://gist.github.com/jackkoppa/0f2b5e8cd683a5261037c685ad8bcc8b)**, after `ng build -prod` and before `http-server`. If you can, give it a shot and see if it works for you? I'm trying to work on a pull request (# 1 above can definitely be fixed in the source, # 2 needs more investigation to avoid breaking anything); currently stymied by setting up a dev environment for the `@angular/service-worker` package

Final note: you can see my app working offline in [this repo](https://github.com/jackkoppa/cityaq), and you can see a clone of that repo from the time where I was having the issue (so it still fails offline) [here](https://github.com/jackkoppa/cityaq-sw-issue)

**Comment by patrickgoeler** [comment]
Hey @jackkoppa, 
your workaround seems to work for me 👍 
I had the same problem that my app would not refresh when offline.
I hope this will be resolved soon..

**Comment by Amirault** [comment]
@jackkoppa it worked for me !

History  : 
Coming from Angular 4.4.4, i upgrade it in 5.2.3 to have full service-worker support.
Everything worked fine but not the urls inside the dataGroup...

Problem :
The cache (dynamic) do not work when offline but my url is reconized by service-workers (Cache storage ok)

Solution :+1:  :
The @jackkoppa script. Since i use it no problem !

Waiting for a fix :+1:

**Comment by KGALLET** [comment]
Gonna test it next week 👍

**Comment by odahcam** [comment]
I'm using the `@angular/service-worker` outside the CLI and I hope it works for me too, because this package never worked in any build made outside CLI, not even a simple `ng eject`. :(

**Comment by ajithere** [comment]
@jackkoppa - It works like a charm!! Thanks a lot. Waiting for the actual fix now

**Comment by MrCroft** [comment]
Not sure if what's causing this HTTP_ERROR 504 also causes this:
![image](https://user-images.githubusercontent.com/2809154/36646026-9e9ec312-1a7a-11e8-87c6-42109f429f97.png)
`Uncaught (in promise) TypeError: Cannot read property 'id' of undefined`
or if it's something new. But sometimes, after I push new code, the app is completely broken. Until I manually unregister the SW. I know that if you delete the `ngsw.json` file, the SW should automatically destroy itself. But, in the above scenario, it doesn't even request that file. I'm not sure deleting it would have any effect 😄 

SW used to "simply work" before 5.x. Oh man! Why did it have to break? 😢 
Luckily this is just a test environment. But, in order to make sure things don't blow up in production, my only option is to remove it before the next deployment. It was fun, SW!

**Comment by lishichao1002** [comment]
same question [https://lishichao1002.github.io/glowworm/](https://lishichao1002.github.io/glowworm/)

**Comment by ZiPengYe** [comment]
@jackkoppa [fix-sw,js](https://gist.github.com/jackkoppa/0f2b5e8cd683a5261037c685ad8bcc8b) not complete work for me,
but I referred to it and made [another one](https://github.com/ZiPengYe/sw-test/blob/master/build/fix-sw.js)

**Comment by markdernie** [comment]
What appears to be happening is that when you go offline, if you then hit reload chrome creates a new client. This renders the app from cache. Within a second or two of that happening ngsw-worker.js sends a request to get ngsw.json (to check whether there is an update to any of the code) this fails (because there is no network). On failing it sets the state of the worker to EXISTING_CLIENTS_ONLY. If you hit reload again then another new client is created and fails to render. This behaviour seems consistent. There are a few hits on google with regard to the problem being related to having dev tools open or that it is a timing issue. There is a timing issue in that if you hit refresh a series of times quickly after you have gone offline the call to ngsw.json doesn't get chance to fail and set EXISTING_CLIENTS_ONLY so the page continues to renders. If you take a deep breath before hitting refresh it will then fail.
If rather than hitting reload, you navigate off the page and back on it works, if your app is more complex and you stay in the app and navigate around it also should also work disconnected. 

If you want to see the current state open a tab and navigate to <base path>/ngsw/state. To reset the state you can stop and start the worker in devtools.

I think the reason this is coming up as a problem is because people are creating a basic app to test this functionality and the easiest way to check is to hit reload after going off line, which then breaks it.

It would seem sensible to allow users to hit reload of an existing client without the sw switching states as I would imagine that would happen a lot, so hopefully this will get revisited. There are some notes in the code about why it works like it does so it looks like this is by design rather than a bug.

**Comment by MrCroft** [comment]
In my opinion, offline means offline. Not "offline, but only if you open a fresh tab, not when hitting refresh". I don't think that's the design, because it would be faulty. Especially since api calls also result in errors and don't work offline (with the app successfully loaded). It's not just the app shell. I don't think that a refresh vs. opening a fresh tab should be treated differently.
As for "having dev tools open" I know it can be an issue. But it behaves the same without dev tools open. Plus, as far as I know, there should be no issues if you're on the network tab (or at least it was the case a couple months ago, don't know if anything changed). When ServiceWorker used to work, the only time it didn't behave as expected was if I had the Application tab active in dev tools, but worked fine while having any other tab active.

**Comment by jackkoppa** [comment]
Thanks @gkalpak - this is fantastic. Appreciate the thorough response, and regret that I wasn't able to do more of the fixing on my own.

I'll test out your fix, to verify that it fixes this issue (21636) as we've been experiencing it, and then, if I still see the `--base-href` problem, open a separate issue. My roadblock up until this point has been building the Angular repo w/ Bazel on Windows ([an example](https://github.com/angular/angular/issues/22508) of that issue). If I can get that working, I expect I could submit a fix for the `--base-href` issue as I've been experiencing it.

**Comment by gkalpak** [comment]
> My roadblock up until this point has been building the Angular repo w/ Bazel on Windows

Me too :grin:
Currently, I still use `sh test.sh browserNoRouter` (but note that certain tests fail - I just ignore those).

**Comment by akilmakda** [comment]
I have also run into the same problem but in my case in local server http-server -p 8000 all images load. but when i upload dist folder to server first time load all images after soft refresh only https images load but http images not load. (when images load form service worker that time issue occurred).

After hard refresh (Ctrl + Shift + r) all images load.

why angular service worker not load http images?

thanks in advance.

**Comment by MrCroft** [comment]
Just tested with Angular 6.0.0-rc.1 and CLI 6.0.0-rc.1
This still happens.
```
> ng new ng6
> cd ng6
> ng g serviceWorker --project ng6
> npm install --save @angular/service-worker@6.0.0-rc.1
> npm run build
> http-server dist/ng6/
```
I did unregister all service workers from localhost (from chrome dev tools) before doing the above steps, so there were no left-overs.

![image](https://user-images.githubusercontent.com/2809154/38372926-a2a41182-38f8-11e8-9ff7-b8bd65d1575b.png)

**Comment by gkalpak** [comment]
@MrCroft, you don't seem to have turned Service Worker on (via the `--service-worker` flag).
Please make sure you have Service Workers turned on and working in the app and if you still have issues with offline mode, open a new issue providing the necessary info (e.g. a minimal reproduction).
Thx!

**Comment by MrCroft** [comment]
@gkalpak The service worker is turned on (I've checked both "angular.json" for `serviceWorker: true` and the generated dist folder for `ngsw-worker.js`) and it is enabled in app.module (`ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production })`).
I think the command `ng g serviceWorker --project ng6` did all that, because I don't remember doing it myself.
Also checked the network tab, when server is running it says "from ServiceWorker".
I also don't think it would be a 504 otherwise.

Anyway, just tried with 6.0.0-rc.2 (both Angular and the CLI), it's still happening but I wasn't sure if I should open another issue, because it's actually the same issue (this one). I thought it should be re-opened rather than re-created.
I opened one now: #23240


P.S. About the `--service-worker` flag, I think it's no longer part of `ng new`. I did try that initially and got no result. Also, not listed under `ng new --help` so that's when I've found `ng g serviceWorker`.
Also, when I tried with rc.1, the build script in package.json was `ng build --prod` that's why I simply ran `npm run build` instead of `ng build --prod` myself (in case anybody's wondering). Now I see the "build" script is back to `ng build`, so with rc.2 I had to run `ng build --prod` myself.

**Comment by jkossis** [comment]
Any updates on the aforementioned issues?

**Comment by alejonext** [comment]
Is the same bug in `6.1.4`.. and well..

**Comment by jhwoodward** [comment]
I have the same issue

**Comment by alejonext** [comment]
The problem is
- Angular compile files (index.html, favicon.ico, etc.)
- Generates the hash
- Generates the `ngsw-config.json` file
- CloudFlare (in my case) takes those files, and modifies them (minifier .. or other things)
- When the angular ServiceWorker checks the files, it is not the same ngsw-config.json hash 
- and FAILS!!! Get a 504.

The solution is very easy!! Use the same minify in Angular, before start the compile!!

**Comment by vallgrenerik** [comment]
I have the same issue. 
504 when offline in Chrome, it works when I try in Firefox. Has anyone else tried in Firefox?

**Comment by vallgrenerik** [comment]
@alejonext Can you elaborate your solution? I don't understand

**Comment by SystemR** [comment]
I can confirm Cloudflare minification affects offline mode. Make sure you purge cache as well when you disable them.

**Comment by david-genger** [comment]
@vallgrenerik Does it work for you now, I am having the same issue. It's working in Firefox (and edge 🤮) but not in chrome

**Comment by vallgrenerik** [comment]
@david-genger no unfortunately not 100% on Chrome on desktop. It works some times and sometimes not.
But on Android phone(and on other desktop browsers as you said) it seems to work.

**Comment by tomfaltesek** [comment]
I'm getting the same issue with Angular 7.

Offline mode works great in Firefox. In Chrome, offline mode gives me a 504. Everything passes in the Chrome Lighthouse tools.

It doesn't work on my Android phone either (Standalone or Chrome browser). Not sure what the issue could be,

**Comment by david-genger** [comment]
I’m voting this issue to be reopened.


________________________________

**Comment by tomfaltesek** [comment]
Quick update: I tried disabling the hash routing and that didn't help. Although, the audit tools still show everything working correctly. Offline mode in Firefox is also working.

Some more details:

1. Open app in chrome
2. Toggle Chrome to offline mode
3. Refresh the page
4. Chrome displays a 504 page. "___ took too long to respond."

The network tab shows a successful request for `ngsw-worker.js`, then a failed fetch request for `ngsw.json?ngsw-cache-bust=0.23549002157983`

**Comment by alejonext** [comment]
@tomfaltesek You read my comment https://github.com/angular/angular/issues/21636#issuecomment-439172813

**Comment by tomfaltesek** [comment]
I figured out my issue, after spending time debugging the ngsw-worker.

During my automated release, I replace some tokens in my "/assets/data/config.json" file. So the hash in the hash table is incorrect once the application is deployed. This causes the `state` to be set to `DriverReadyState.EXISTING_CLIENTS_ONLY` and the file is never recovered from cache.

As a temporary workaround, I've commented out `this.state = DriverReadyState.EXISTING_CLIENTS_ONLY;` in the ngsw-worker file. Now offline mode works great.

**Comment by alejonext** [comment]
@tomfaltesek and @david-genger 
Well one a simple modification in one file, in the list of `ngsw.json`  thats is all with fails the SW for angular. Check the files that are coming out in your domain, and how they were compiled by angular. Remember that a single character can make a difference

**Comment by tomfaltesek** [comment]
@alejonext - There's nothing to check.

The config.json that is served to the browser is surely different than the config.json the existed during build time. As I said, during our release process, many strings are replaced in that file (depending on the env to which you are releasing). So it's not a single character; it's many characters. I've intentionally changed the file after build.

**Comment by deniszholob** [comment]
I have the same issue ad Tom and others.
Offline in Firefox is ok on desktop. On android FF refresh shows a blank page, but a browse to the page in Offline mode works.
In Chrome, it just shows the 504 on desktop or android.

My code is available here: https://github.com/deniszholob/angular-app-template

![chrome_2019-02-21_19-01-51](https://user-images.githubusercontent.com/3830741/53214819-4a43ac80-360b-11e9-98f9-c66d40ec5920.png)

==========

Now i did get it to work locally, using http-server but only after I hit refresh a couple times, then go offline. (while using incognito mode, why the requirement for a couple refreshes then???)

1st offline refresh, this time actually only didnt load images

![chrome_2019-02-21_18-58-06](https://user-images.githubusercontent.com/3830741/53214727-eae59c80-360a-11e9-90fe-9dbb18d28eeb.png)

Then after going live and hitting a refresh a couple times then going offline and hitting refresh only thing that didnt load is font awesome.
![chrome_2019-02-21_18-58-29](https://user-images.githubusercontent.com/3830741/53214739-f638c800-360a-11e9-844e-3a4e33141d34.png)

**Comment by Taylorsuk** [comment]
Did anyone get this fixed? I'm having this issue in Apr 19 on Angular 7.

**Comment by DanielRuf** [comment]
We should reopen this one.

**Comment by alejonext** [comment]
Why repoen?

1. READ my comment https://github.com/angular/angular/issues/21636#issuecomment-435404024
2. Search the problem in the CNAME or your provider

**Comment by DanielRuf** [comment]
We still see the issue in the latest release of Angular 7 in specific browsers. Others reported this too (see the previous replies).

**Comment by DanielRuf** [comment]
As context for you, we use the same commands locally and on CI and see this without any proxy or service which changes the files. @diverent2 can tell you more.

**Comment by alejonext** [comment]
@DanielRuf That you have services from third  that modify your files, is not a problem of Angular, it is a problem of your third. Well, because you are modifying the files generated by the CLI. And even so, there are many methods to use Service Workers, not just Angular.  Given that is not a Bug

**Comment by DanielRuf** [comment]
We have no thirdparty which changes the files in our case ;-) Our CI builds rsync the files to the server and server content dies not change.

**Comment by RamBoFe** [comment]
Hello,
I had the same problem and after reading the conversation, it turns out that my problem was with the FTP client I was using to send my Angular project to the host.
I was using FileZilla and noticed that the size of the files sent including the ngsw.json file was different from my local file once on the FTP server. And that was what caused my problem. Since I use WinSCP I have no problem and my project works offline :)

**Comment by DanielRuf** [comment]
Hi @RamBoFe,

you might like to try CyberDuck too. The problem my be normal vs binary transfer (use the binary one).

**Comment by RamBoFe** [comment]
Thank you @DanielRuf, 
I'll try CyberDuck ;)

**Comment by danieldanielecki** [comment]
Same issue here, working on Firefox "Work Offline", but on Chrome have 504. Using Angular 7 with Firebase hosting.

**Comment by david-genger** [comment]
Same here using chrome 74 and Angular 8. Not working.

**Comment by zain-b** [comment]
Same issue, Angular 8

**Comment by jrcasso** [comment]
Same issue, Angular 8 also. Lazy-loaded feature modules, the works.

**Comment by zuschy** [comment]
Thanks for this comment! For a long time, I thought, I'm the only one who feels like that!

**Comment by H--o-l** [comment]
@jrcasso @zuschy guys (or girl) it's an open-source project.

First, thank people that have given time building it so you do not have to do it yourself.

Second, if you think it should be improved, and have idea how to do it, open issues/PRs.
You have the right to look at the service code, it's not that hard.

Finally, if this code does not match your needs, there is other pre-written service workers out there which work fine with Angular.

I personnally think the Angular service worker have some unique concepts that makes it powerfull (no need to `skipWait()` between app versions, multiple app versions in parrallel, nice self-hosted debug page `/ngsw/state`, in-app service to control it, etc), but you have to be meticulous at configuring it.

If there are bugs, your comments certainly does not help fixing them.

**Comment by zuschy** [comment]
@H--o-l don't get me wrong, I agree with you.

But sometimes it appears, that the service worker arbitrary does or does't cache requests.
I built the same code, serveral times, on different servers, but the behavior is never the same. I try continuously to find the right configuration, but the behavior of the service worker isn't clearly reproducible, this is frustrating.
Sometimes it's good to know, that you're not alone with a problem like that.

That doesn't mean, that we does't try to find a way to fix the bug! ;)

**Comment by MrCroft** [comment]
@H--o-l 
To be honest, I haven't checked lately if the same reproduction steps still apply, or if people are running into this in some more specific scenarios. I guess it doesn't bother me that much because I don't have any apps that truly need to work offline, I only use the ServiceWorker to benefit from faster load times (after that first load, when it builds the cache, of course).

>  if you think it should be improved, and have idea how to do it, open issues/PRs.
You have the right to look at the service code, it's not that hard.  

As simple as it may look for some, other people surely either don't have the time or the understanding of what's going on there. I tend to understand people's frustration, if an issue has been marked as resolved but it's still happening for them. As long as they're not hateful, I usually "thumbs up" comments that report the issue still happening. I always thought that reopening an issue is better than opening a new one, identical to an existing one. But that's just my opinion.

> if this code does not match your needs, there is other pre-written service workers out there which work fine with Angular.

I don't think this is about "needs matching". As long as the app loads in offline mode when accessing `/` and doesn't load when accessing `/some/route/` (and we're not talking about a misconfigured server - especially if we're "offline", that wouldn't even matter, since we don't even reach the server), that seems like a bug. If I really needed my apps to "fully" work offline, I guess this would be frustrating for me as well, to have the issue resolved/closed, but the reported behavior still happening.

> I personally think the Angular service worker have some unique concepts that makes it powerfull [...]

Agree! I guess that's why people prefer the built-in option, rather than 3rd party solutions.
But, like I said, not everybody has the time or understanding to be able to come with bug fixes, improvements etc. I don't think that they don't appreciate the team, the contributors, the community. Maybe just lacking expressing gratitude.

> First, thank people that have given time building it so you do not have to do it yourself.

That being said, I always thought that big "thank you" 's are implied 😃. Surely, every developer using it is grateful for everything that the Angular framework (or, rather, the entire ecosystem) does for us, grateful for everything that the core team and every contributor is doing, and for the whole community. But I guess you're right, not saying it makes it seem like it's not appreciated enough.

**Comment by jrcasso** [comment]
@H--o-l 

> First, thank people that have given time building it so you do not have to do it yourself.

You're confusing my frustration with ingratitude.

> Second, if you think it should be improved, and have idea how to do it, open issues/PRs.
> You have the right to look at the service code, it's not that hard.

This is essentially the same as saying "You fix it." which I did, outside of the framework. I don't know what kind of life you have where you have the time to PR Angular issues, but I don't have it. Frankly, that's the reason I picked Angular as a framework in the first place; I figured it was reliable and highly functional requiring minimal non-native additions. This is clearly not the case with Service Workers, and I disagree with your assertion that my comment isn't helpful. It's a closed issue that I believe to be very open, and voicing my frustration about that can bring attention to the issue. Moreover, I provided a small snippet that fails in my use-case.

> Finally, if this code does not match your needs, there is other pre-written service workers out there > which work fine with Angular.

Bingo. If life affords me some time, I'll see if my fix applies to Angular and submit a PR.

> I personnally think the Angular service worker have some unique concepts that makes it powerfull (no need to skipWait() between app versions, multiple app versions in parrallel, nice self-hosted debug page /ngsw/state, in-app service to control it, etc), but you have to be meticulous at configuring it.
> 
> If there are bugs, your comments certainly does not help fixing them.

Obviously I agree Angular is great. That's why we're here in the first place. But I also don't think your comment helped the issue. In fact, it seems like you focused entirely on the tone of my comment than the content: _the issue is still open_.

You might notice more frustration in the tone here; that's because I find your fixation on my tone in the first comment to be unnecessary in the context of fixing the issue.
